GCE OSLogin module: init

adisbladis opened the pull request:
###### Motivation for this change
###### TODO
 - [x] get  #50316 merged, rebase on master
 - [x] fix sudo
 - [x] integrate with gke profiles
 - [x] add to release notes

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

zimbatm commented on nixos/modules/security/pam.nix:
```diff
@@ -77,6 +77,15 @@ let
+        type = types.bool;
+        description = ''
+          If set, the Google OS Login PAM module will be used.
+          This only works Google Compute Engine instances.
```
```suggestion
          This only works on Google Compute Engine instances.
```

flokli commented:
cc @arianvp I cherry-picked commits from #50316 in here as well.

flokli commented on nixos/modules/config/nsswitch.nix:
```diff
@@ -15,6 +15,7 @@ let
   ldap = canLoadExternalModules && (config.users.ldap.enable && config.users.ldap.nsswitch);
   sssd = canLoadExternalModules && config.services.sssd.enable;
   resolved = canLoadExternalModules && config.services.resolved.enable;
+  googleOSLogin = canLoadExternalModules && config.security.googleOSLogin.enable;
```
I'm not sure on why some modules seem to set this here, and some from their module itself. Any pointers?

zimbatm replied:
No idea why. In the case of googleOSLogin it would be better if this could be set by the module itself as it's very target-specific.

adisbladis replied:
The main reason I chose to set this here is that `hostArray` cant be modified externally right now.
If we added facilities for that I'd like it if each individual module took care of it's own `nss` config.

flokli replied:
@flokli You meant `passwdArray`, and in this case, it's like all other modules are currently doing it too. In some cases, order matters here, so it makes sense to do it inside the nsswitch module.

flokli commented on nixos/modules/security/google_oslogin.nix:
```diff
@@ -0,0 +1,63 @@
+      #includedir /run/google-sudoers.d
+    '';
+    systemd.tmpfiles.rules = [
+      "d /run/google-sudoers.d 660 root root -"
```
The `-` for age means systemd will create the directory with mentioned permissions and ownership, but not automatically clean up anything.

I moved that from `/var/google-sudoers.d` (and patched `pam_module/pam_oslogin_admin.cc`), as that location seemed much more suitable for runtime files.

`pam_oslogin_admin.cc` has code to clean up existing files if admin status is revoked, it's just a more meaningful location (and cleans up expired users on a reboot)

flokli commented on nixos/modules/security/google_oslogin.nix:
```diff
@@ -0,0 +1,63 @@
+    ];
+
+    # enable the nss module, so user lookups etc. work
+    system.nssModules = [ package ];
```
see above, either set it here, or in https://github.com/NixOS/nixpkgs/pull/51566/files#diff-5796c52b71eee35842f408f4126430d6R103

flokli commented on nixos/modules/security/google_oslogin.nix:
```diff
@@ -0,0 +1,63 @@
+    system.nssModules = [ package ];
+
+    # Ugly: sshd refuses to start if a store path is given because /nix/store is group-writable.
+    # So indirect by a symlink.
```
This hack is borrowed from `nixos/modules/services/misc/sssd.nix`.

flokli commented on nixos/modules/security/pam.nix:
```diff
@@ -193,7 +202,8 @@ let
       makeHomeDir = mkOption {
-        default = false;
+        # this needs to be enabled if GoogleOSLogin is enabled
+        default = config.security.pam.enableGoogleOSLogin;
```
This affects all `security.pam.services.<name?>.makeHomeDir`.

Do we want to do that only for logins over ssh, not when switching to that user via different means?
Does that even make a difference, or is it more of a philosophical question?

adisbladis replied:
Honestly I'm not sure. I had the same thought as you but dismissed it as I'd think it was more surprising if things didn't work the same across services.

flokli replied:
This was changed.

Setting `security.googleOsLogin.enable` to true will now set
```
security.pam.services.sshd.makeHomeDir = true
security.pam.services.sshd.googleOsLoginAccountVerification = true
```

And we'll add a trigger for setting `googleOsLoginAuthentication`. This removes some oslogin specific cruft from the PAM module itself, and only configures oslogin pam for sshd.

flokli commented on nixos/modules/security/pam.nix:
```diff
@@ -279,6 +289,11 @@ let
               "account sufficient ${pam_krb5}/lib/security/pam_krb5.so"}
 
+          ${optionalString cfg.googleOsLoginAuth ''
+            account optional ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_admin.so
```
Rules borrowed from [google_oslogin_control](https://github.com/GoogleCloudPlatform/compute-image-packages/blob/master/google_compute_engine_oslogin/bin/google_oslogin_control) - somebody deeper into pam should definitely take a look on that.

zimbatm replied:
Is it possible to set this from the googleOSLogin module directly instead. Ideally the PAM module would provide the interface to allow googleOSLogin to set the right settings from the outside.

flokli replied:
Order is important here I think.

Getting that right is not very easy to do with the current module system. There is a TODO since https://github.com/NixOS/nixpkgs/commit/39bffdb34c1e1f344ba350041c10d0b07deeecb0 (from 2009) to provide such hooks. I see that outside the scope of this PR.

flokli replied:
I did dig deeper, and also checked back in https://github.com/GoogleCloudPlatform/compute-image-packages/issues/717.

As described there, we add the `oslogin_login.so` and `oslogin_admin.so` modules for the auth account management group at the end, and `oslogin_login.so` auth module at the beginning, with recommended upstream parameters.

zimbatm commented on nixos/doc/manual/release-notes/rl-1903.xml:
```diff
@@ -245,6 +245,66 @@
+     <literal>libnss_systemd.so</literal> module which is used by
+     <literal>systemd</literal> to manage uids and usernames in the presence
+     of <literal>DynamicUser=</literal> in systemd services.
+     The was already the default behaviour in presence of
```
```suggestion
     This was already the default behaviour in presence of
```

flokli replied:
reported on https://github.com/NixOS/nixpkgs/pull/50316

zimbatm commented on nixos/doc/manual/release-notes/rl-1903.xml:
```diff
@@ -245,6 +245,66 @@
+     of <literal>DynamicUser=</literal> in systemd services.
+     The was already the default behaviour in presence of
+     <literal>services.sssd.enable = true</literal> because nscd caching
+     would interfere sssd in unpredictable ways as well.Because we're using nscd
```
```suggestion
     would interfere with sssd in unpredictable ways as well. Because we are using nscd
```

flokli replied:
reported on https://github.com/NixOS/nixpkgs/pull/50316

flokli commented:
cc @Assassinkin @matthewbauer for other recent pam contributions

adisbladis commented:
@flokli Could you squash the fixup commits? I don't mind you force-pushing to my fork.

flokli commented:
I'd still like to get the TODOs fixed and #50316 merged before rebasing again.

Any ideas about the " PAM unable to resolve symbol:" errors?

flokli commented:
Ok, after some deeper look into pam, this is due to us misusing some oslogin modules in wrong management groups. I'll update this accordingly.

flokli commented on nixos/modules/security/pam.nix:
```diff
@@ -77,6 +77,15 @@ let
       };
 
+      googleOsLoginAuth = mkOption {
+        default = config.security.pam.enableGoogleOsLogin;
```
This does only make sense for `/etc/pam.d/sshd`, so better make that default to false and let `security.googleOSLogin.enable` enable `security.pam.services.sshd.googleOsLoginAuth`.

flokli replied:
This was changed.

Setting `security.googleOsLogin.enable` to true will now set
```
security.pam.services.sshd.makeHomeDir = true
security.pam.services.sshd.googleOsLoginAccountVerification = true
```

And we'll add a trigger for setting `googleOsLoginAuthentication`. This removes some oslogin specific cruft from the PAM module itself, and only configures oslogin pam for sshd.

flokli committed:
google-compute-engine-oslogin: init at 1.4.3

flokli committed:
security.pam.services.<name?>.: add googleOsLogin(AccountVerification|Authentication)

flokli committed:
config.security.googleOsLogin: add module

The OS Login package enables the following components:
AuthorizedKeysCommand to query valid SSH keys from the user's OS Login
profile during ssh authentication phase.
NSS Module to provide user and group information
PAM Module for the sshd service, providing authorization and
authentication support, allowing the system to use data stored in
Google Cloud IAM permissions to control both, the ability to log into
an instance, and to perform operations as root (sudo).

flokli committed:
config.nsswitch: load cache_oslogin and oslogin nss modules if config.security.googleOsLogin.enable is set

flokli committed:
nixos/tests: add google-oslogin test

flokli commented:
Finally got sudo working - culprit was the `pam_unix.so` module configured wrongly in NixOS (see df7b1eb11afd8429018e2a1eea95e452610a7e14) for a detailed explanation.

This was fixed for sssd only in https://github.com/NixOS/nixpkgs/pull/31969, @dezgeg, @PsyanticY, can you have a look at df7b1eb11afd8429018e2a1eea95e452610a7e14, too?

flokli committed:
nixos/modules/virtualisation/google-compute-config.nix: remove google-accounts-daemon

Use googleOsLogin for login instead.
This allows setting users.mutableUsers back to false, and to strip the
security.sudo.extraConfig.

security.sudo.enable is default anyhow, so we can remove that as well.

adisbladis commented:
Things seems to be working great now! Thanks @flokli for all the good work.

PsyanticY commented:
@flokli I think it should be like that for all module not just sssd. `account required pam_unix.so`.  when i fixed that for SSSD I mentioned that it should be like this for all modules but i thought maybe there was some use case for it being `sufficient` in nixos even thought all the other disto use `required`

PsyanticY commented:
Even this part 
```
          ${optionalString (config.services.sssd.enable && cfg.sssdStrictAccess==false)
               "account sufficient ${pkgs.sssd}/lib/security/pam_sss.so"}
           ${optionalString (config.services.sssd.enable && cfg.sssdStrictAccess)
              "account [default=bad success=ok user_unknown=ignore] ${pkgs.sssd}/lib/security/pam_sss.so"}
```
it should not be sufficient. I would suggest removing the `sssdStrictAccess` option and swapping sufficient with this ` [default=bad success=ok user_unknown=ignore] ` .

flokli commented:
Reading up on https://wiki.debian.org/LDAP/PAM, it seems this should at least work if all these 'external' pam modules provide an nss module, too. There are other examples on what can/should be done instead - but it seems to be very a combinatory hell.

Maybe we should limit the pam module to only allow one external pam module (sssd/ldap/oslogin/kerberos) to be active, and check configuration for them?

flokli committed:
release-notes/rl-1903: add security.googleOsLogin

flokli commented:
With https://github.com/NixOS/nixpkgs/pull/52488 being merged, this should be good to go, too.
